File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/PkeySqlAdapter.java
package app.cash.backfila.client.misk.base;


import misk.hibernate.Id;

public class PkeySqlAdapter {
    public Object fromString(Class<?> type, String sqlString) {
        if (type == String.class) return sqlString;
        if (type == Id.class) return new Id(Long.parseLong(sqlString));
        throw new IllegalArgumentException(
                String.format("Unsupported backfill primary key type: %s. Add an adapter to this class or "
                        + "use a different type as your pkey.", type));
    }
}

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/PkeySqlAdapter.kt
package app.cash.backfila.client.misk

import misk.hibernate.DbEntity
import misk.hibernate.Id
import javax.inject.Inject

class PkeySqlAdapter @Inject constructor() {
  @Suppress("UNCHECKED_CAST") // We use runtime checks to guarantee casts are safe.
  fun <Pkey> pkeyFromString(type: Class<Pkey>, sqlString: String): Pkey {
    return when (type) {
      String::class.java -> sqlString as Pkey
      Id::class.java -> Id<DbPlaceholder>(sqlString.toLong()) as Pkey
      else -> throw IllegalArgumentException(
          "Unsupported backfill primary key type: $type. Add an adapter to this class or "
              + "use a different type as your pkey.")
    }
  }

  /** This placeholder exists so we can create an Id<*>() without a type parameter. */
  private class DbPlaceholder : DbEntity<DbPlaceholder> {
    override val id: Id<DbPlaceholder> get() = error("unreachable")
  }
}